### PR TITLE
Fix azure_rm_acs due to Azure breaking API change

### LINKED ISSUE #32526: azure_rm_acs fails due to breaking API changes
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME
azure_rm_acs

##### ANSIBLE VERSION
2.5.0

##### CONFIGURATION
NA

##### OS / ENVIRONMENT
NA

##### SUMMARY
Due to breaking changes to the Azure Python SDK in `azure-mgmt-containerservice>=2.0.0`, we'll need to adjust the code and requirements to continue working properly. More details in https://github.com/ansible/ansible/pull/32518

##### STEPS TO REPRODUCE
Install/upgrade `azure-mgmt-containerservice` to >=2.0.0 and attempt to use azure_rm_acs module

##### EXPECTED RESULTS
success

##### ACTUAL RESULTS
exception (see #32518)

### LINKED ISSUE #33120: azure_rm_functionapp module doesn't work
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

azure_rm_functionapp

##### ANSIBLE VERSION

```
ansible 2.5.0
  config file = None
  configured module search path = [u'/Users/zhijzhao/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Sep 25 2017, 09:53:22) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)]
```

##### CONFIGURATION

No change at all.

##### OS / ENVIRONMENT

macOS 10.13.1

##### SUMMARY

Exceptions as below when creating azure function.

TASK [create basic function app] ************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'AzureRMFunctionApp' object has no attribute 'base_url'
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/0r/_5p333ds2q30tf1pcq6p_wc40000gn/T/ansible_Aun21J/ansible_module_azure_rm_functionapp.py\", line 297, in <module>\n    main()\n  File \"/var/folders/0r/_5p333ds2q30tf1pcq6p_wc40000gn/T/ansible_Aun21J/ansible_module_azure_rm_functionapp.py\", line 294, in main\n    AzureRMFunctionApp()\n  File \"/var/folders/0r/_5p333ds2q30tf1pcq6p_wc40000gn/T/ansible_Aun21J/ansible_module_azure_rm_functionapp.py\", line 156, in __init__\n    required_if=required_if\n  File \"/var/folders/0r/_5p333ds2q30tf1pcq6p_wc40000gn/T/ansible_Aun21J/ansible_modlib.zip/ansible/module_utils/azure_rm_common.py\", line 303, in __init__\n  File \"/var/folders/0r/_5p333ds2q30tf1pcq6p_wc40000gn/T/ansible_Aun21J/ansible_module_azure_rm_functionapp.py\", line 174, in exec_module\n    function_app = self.web_client.web_apps.get(\n  File \"/var/folders/0r/_5p333ds2q30tf1pcq6p_wc40000gn/T/ansible_Aun21J/ansible_modlib.zip/ansible/module_utils/azure_rm_common.py\", line 783, in web_client\nAttributeError: 'AzureRMFunctionApp' object has no attribute 'base_url'\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 0}

##### STEPS TO REPRODUCE

```yaml
---
- hosts: localhost
  tasks:
  - name: create storage account for function apps
    azure_rm_storageaccount:
      resource_group: zhijzhao
      name: azfunccistor4
      account_type: Standard_LRS

  - name: create basic function app
    azure_rm_functionapp:
      resource_group: zhijzhao
      name: azfuncci
      storage_account: azfunccistor4

```

##### ACTUAL RESULTS

```

```

### PR DESCRIPTION
**Pull request by ZhijunZhao**
##### SUMMARY

Fix `azure_rm_acs` due to Azure breaking API change. 

This PR also includes minor cleanup that can fix the breaking module `azure_rm_functionapp`.

Fix #32526
Fix #33120 

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

azure_rm_acs
azure_rm_functionapp

##### ANSIBLE VERSION

```
ansible 2.5.0
  config file = None
  configured module search path = [u'/Users/zhijzhao/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Sep 25 2017, 09:53:22) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)]
```

##### ADDITIONAL INFORMATION

```

```

@xscript @yuwzho @zikalino @yaweiw

### CODE REVIEW

File: lib/ansible/modules/cloud/azure/azure_rm_acs.py [hunk 1]
```diff
@@ -325,8 +330,10 @@ def create_master_profile_instance(masterprofile):
     '''
     return ContainerServiceMasterProfile(
         count=masterprofile[0]['count'],
-        dns_prefix=masterprofile[0]['dns_prefix']
-    )
+        dns_prefix=masterprofile[0]['dns_prefix'],
+        vm_size=masterprofile[0]['vm_size'],
+        first_consecutive_static_ip=None
```
Comment (ZhijunZhao) [inline lib/ansible/modules/cloud/azure/azure_rm_acs.py]: So far, Azure server doesn't accept request with body that contains `first_consecutive_static_ip` property.

Comment (xscript) [reply lib/ansible/modules/cloud/azure/azure_rm_acs.py]: should add some comments to explain why we always want to set it to none.

File: lib/ansible/modules/cloud/azure/azure_rm_acs.py [hunk 2]
```diff
@@ -62,6 +62,11 @@
                   - 1
                   - 3
                   - 5
+            vm_size:
+                description:
+                    - The VM Size of each of the Agent Pool VM's (e.g. Standard_F1 / Standard_D2v2).
+                required: true
+                default: Standard_D2v2
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_acs.py]: version_added: "2.5"

File: lib/ansible/modules/cloud/azure/azure_rm_acs.py [hunk 3]
```diff
@@ -564,12 +583,12 @@ def exec_module(self, **kwargs):
                         to_be_updated = True
 
                     # Cannot Update the master count for now // Uncomment this block in the future to support it
-                    if response['master_profile'].get('count') != self.master_profile[0].get('count'):
+                    if response['master_profile'].get('count') != self.master_profile[0].get('count') or response['master_profile'].get('vm_size') != self.master_profile[0].get('vm_size'):
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_acs.py]: Why not make the logic handled by Azure service

File: lib/ansible/modules/cloud/azure/azure_rm_acs.py [hunk 4]
```diff
@@ -564,12 +583,12 @@ def exec_module(self, **kwargs):
...
                         # self.log(("Master Profile Count Diff, Was {0} / Now {1}"
                         #           .format(response['master_profile'].count,
                         #           self.master_profile[0].get('count'))))
                         # to_be_updated = True
-                        self.module.warn("master_profile.count cannot be updated")
+                        self.module.warn("master_profile.count or master_profile.vm_size cannot be updated")
```
Comment (xscript) [inline lib/ansible/modules/cloud/azure/azure_rm_acs.py]: should extract a method doing the check to avoid duplicate code

**Comment by xscript** [comment]
shipit

**Comment by yuwzho** [comment]
shipit

File: lib/ansible/modules/cloud/azure/azure_rm_acs.py [hunk 5]
```diff
@@ -62,6 +61,12 @@
                   - 1
                   - 3
                   - 5
+            vm_size:
+                description:
+                    - The VM Size of each of the Agent Pool VM's (e.g. Standard_F1 / Standard_D2v2).
+                required: true
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_acs.py:65]: If there is a default, why would `required: true` be set?

File: lib/ansible/modules/cloud/azure/azure_rm_acs.py [hunk 6]
```diff
@@ -393,12 +401,24 @@ def create_master_profile_dict(masterprofile):
...
+
+def create_service_principal_profile_dict(serviceprincipalprofile):
+    '''
+    Helper method to deserialize a ContainerServiceServicePrincipalProfile to a dict
+    :param: serviceprincipalprofile: ContainerServiceServicePrincipalProfile with the Azure callback object
+    :return: dict with the state on Azure
+    '''
+    return dict(
+        client_id=serviceprincipalprofile.client_id,
+        secret=serviceprincipalprofile.secret
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_acs.py]: I wouldn't return this value, if it is a secret then we don't want to expose it.

**Review by jborean93 — COMMENTED** [review]
Just a few minor questions I have and some changes. I'm not sure on the packaging side so I might let @nitzmahone comment on that.

File: lib/ansible/modules/cloud/azure/azure_rm_acs.py (hunk 5 above)
Comment (ZhijunZhao) [reply lib/ansible/modules/cloud/azure/azure_rm_acs.py:65]: Good spot!

File: lib/ansible/modules/cloud/azure/azure_rm_acs.py (hunk 6 above)
Comment (ZhijunZhao) [reply lib/ansible/modules/cloud/azure/azure_rm_acs.py]: You're quite right on this. Actually the secret in response json is empty. Let me remove the secret key from the dict.

**Comment by ZhijunZhao** [comment]
@jborean93  @nitzmahone Regarding the packaging, I'm just following how the versions of other azure modules are handled. I think it's to avoid breaking api change.

**Comment by ZhijunZhao** [comment]
ready_for_review

**Review by nitzmahone — APPROVED** [review]

**Pull request merged**